sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/unit/read/SnowflakeIOReadTest.java

```diff
@@ -0,0 +1,251 @@
+  public static final String FAKE_TABLE = "FAKE_TABLE";
+
+  @Rule public final transient TestPipeline pipeline = TestPipeline.create();
+  @Rule public ExpectedException exceptionRule = ExpectedException.none();
+
+  private static SnowflakeIO.DataSourceConfiguration dataSourceConfiguration;
+  private static BatchTestPipelineOptions options;
+
+  private static SnowflakeService snowflakeService;
+  private static SnowflakeCloudProvider snowflakeCloudProvider;
+
+  private static String stagingBucketName;
+  private static String integrationName;
+
+  private static List<GenericRecord> avroTestData;
+
+  @BeforeClass
+  public static void setup() {
+
+    List<String> testData = Arrays.asList("Paul,51,red", "Jackson,41,green");
+
+    avroTestData =
+        ImmutableList.of(
+            new AvroGeneratedUser("Paul", 51, "red"),
+            new AvroGeneratedUser("Jackson", 41, "green"));
+
+    FakeSnowflakeDatabase.createTableWithElements(FAKE_TABLE, testData);
+
+    PipelineOptionsFactory.register(BatchTestPipelineOptions.class);
+    options = TestPipeline.testingPipelineOptions().as(BatchTestPipelineOptions.class);
+    options.setServerName("NULL.snowflakecomputing.com");
+    options.setStorageIntegration("STORAGE_INTEGRATION");
+    options.setStagingBucketName("BUCKET");
+
+    stagingBucketName = options.getStagingBucketName();
+    integrationName = options.getStorageIntegration();
+
+    dataSourceConfiguration =
+        SnowflakeIO.DataSourceConfiguration.create(new FakeSnowflakeBasicDataSource())
+            .withServerName(options.getServerName());
+
+    snowflakeService = new FakeSnowflakeServiceImpl();
+    snowflakeCloudProvider = new FakeSnowflakeCloudProvider();
+  }
+
+  @Test
+  public void testConfigIsMissingStagingBucketName() {
+    exceptionRule.expect(IllegalArgumentException.class);
+    exceptionRule.expectMessage("withStagingBucketName() is required");
+
+    SnowflakeIO.Read<GenericRecord> read =
+        SnowflakeIO.<GenericRecord>read(snowflakeService, snowflakeCloudProvider)
+            .withDataSourceConfiguration(dataSourceConfiguration)
+            .fromTable(FAKE_TABLE)
+            .withIntegrationName(integrationName)
+            .withCsvMapper(getCsvMapper())
+            .withCoder(AvroCoder.of(AvroGeneratedUser.getClassSchema()));
+
+    read.expand(null);
+  }
+
+  @Test
+  public void testConfigIsMissingIntegrationName() {
+    exceptionRule.expect(IllegalArgumentException.class);
+    exceptionRule.expectMessage("withIntegrationName() is required");
+
+    SnowflakeIO.Read<GenericRecord> read =
+        SnowflakeIO.<GenericRecord>read(snowflakeService, snowflakeCloudProvider)
+            .withDataSourceConfiguration(dataSourceConfiguration)
+            .fromTable(FAKE_TABLE)
+            .withStagingBucketName(stagingBucketName)
+            .withCsvMapper(getCsvMapper())
+            .withCoder(AvroCoder.of(AvroGeneratedUser.getClassSchema()));
+
+    read.expand(null);
```

Comment:
Can we introduce tests to actually read data (from SnowFlakeFakeServiceImpl) and verify results. Please see DirectRunner-based BQ pipeline read tests.

https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOReadTest.java#L285

Also there's no need to manually call 'expand' in tests. We can just execute regular Beam pipelines that will perform the expansion.

Reply:
Thank you @chamikaramj for pointing this. I changed  this tests so now it uses `pipeline.run()` instead of `expand`. The interesting thing is that I had to use `TestRule` similarly as in BQ tests (otherwise tests were failing with `Outputs for non-root node SnowflakeIO.Read are null`), do you know why is that? Is it something connected with "clearing" the pipeline for each test?

In case of reading data by `SnowFlakeFakeServiceImpl` there is a test `testReadWithConfigIsProper` (previously `testConfigIsProper`) [L248](https://github.com/apache/beam/pull/11360/commits/c5abe5ab21d040e45ab75f95dd93cfa1490b4bf9#diff-8b70435e69bb0a87cf801ea7c4ad3c4e) which does that. Do you think there should be more tests like this?